src/core/public/application/application_service.tsx

```diff
@@ -145,6 +145,23 @@ export class ApplicationService {
       this.subscriptions.push(subscription);
     };
 
+    const wrapMount = (plugin: PluginOpaqueId, app: App<any>): AppMount => {
+      let handler: AppMount;
+      if (isAppMountDeprecated(app.mount)) {
+        handler = this.mountContext!.createHandler(plugin, app.mount);
+        // eslint-disable-next-line no-console
+        console.warn(
+          `App [${app.id}] is using deprecated mount context. Use core.getStartServices() instead.`
+        );
+      } else {
+        handler = app.mount;
+      }
+      return async params => {
+        this.currentAppId$.next(app.id);
+        return handler(params);
+      };
```

Comment:
We had an inconsistency on when we were firing the new `currentAppId$` value:

When navigating using `history.push`, it was fired from the wrapped `mount` handler itself, which was awaiting the completion of the app handler, meaning that is was firing after the app finished mounting.

When navigating using the `navigateToApp` API, we were firing it just after performing history navigation, meaning that it did not wait the app to be mounted.

https://github.com/elastic/kibana/blob/78237b2525389f1f1c8a131164a9a0be71fef6c6/src/core/public/application/application_service.tsx#L285-L294

I'm not even sure why we have this `this.currentAppId$.next(appId);` line in `navigateToApp`, as the navigation will change the route, that will call the wrapped mount function that will do the same thing anyway. However removing it was breaking a few tests, so I kept it for now.

For consistency (and because it helped improving this fix), I changed the logic of the wrapped `mount` handler to emit before executing the actual app `mount`.